### PR TITLE
Don't hardcode @cookbook_name as the root_dir name of a cookbook in get_filename

### PR DESCRIPTION
**Pull request by chuhn**
Redone a part of the original patch for https://tickets.opscode.com/browse/CHEF-3307 from chef/chef@5713a00

Just works.

### CODE REVIEW

**Comment by coderanger** [comment]
Can you go in to more detail? As you can see in the delta, we're already (correctly) using the cookbook name here. What does this fix today?

File: lib/chef/cookbook/file_system_file_vendor.rb [hunk 1]
```diff
@@ -40,14 +40,17 @@ def initialize(manifest, *repo_paths)
...
       def get_filename(filename)
-        location = @repo_paths.inject(nil) do |memo, basepath|
-          candidate_location = File.join(basepath, @cookbook_name, filename)
-          memo = candidate_location if File.exist?(candidate_location)
-          memo
+        if cookbooks.has_key?(@cookbook_name)
+          location = File.join(cookbooks[@cookbook_name].root_dir, filename)
+          location = nil unless File.exist?(location)
         end
         raise "File #{filename} does not exist for cookbook #{@cookbook_name}" unless location
```
Comment (lamont-granquist) [inline lib/chef/cookbook/file_system_file_vendor.rb]: the logic could be cleaned up to be a little less tortured:

```ruby
def get_filename(filename)
  location = File.join(cookbooks[@cookbook_name].root_dir, filename) if cookbooks.has_key?(@cookbook_name)
  raise "File #{filename} does not exist for cookbook #{@cookbook_name}" unless location && File.exist?(location)
```

**Comment by lamont-granquist** [comment]
needs a spec test, and TL;DR for how it addresses CHEF-3307 would be useful

**Comment by chuhn** [comment]
I'm not really accustomed to the Github issue tracker, sorry. This patch fixes #5089. 
The issue is described there.

**Comment by chuhn** [comment]
Short summary:
You must not assume that `cookbooks[@cookbook_name].root_dir == File.join(basepath, @cookbook_name)`.
If the directory name cannot be different from the cookbook name, a mandatory name attribute in `metadata.rb`makes no sense.

**Comment by coderanger** [comment]
I'm not sure that can happen anymore though, maybe with solo in legacy mode?

**Comment by chuhn** [comment]
It happens with chefspec and template tests. Please read #5089.

**Comment by lamont-granquist** [comment]
- needs the logic cleaned up
- needs a spec

**Comment by chef-supermarket** [comment]
Hi. I am an automated pull request bot named Curry. There are commits in this pull request whose authors are not yet authorized to contribute to Chef Software, Inc. projects or are using a non-GitHub verified email address. To become authorized to contribute, you will need to sign the Contributor License Agreement (CLA) as an individual or on behalf of your company. [You can read more on Chef's blog.](https://www.getchef.com/blog/2014/06/23/changes-to-the-contributor-license-agreement-process)

## GitHub Users Who Are Not Authorized To Contribute

The following GitHub users do not appear to have signed a CLA:

* @chuhn

[Please sign the CLA here.](https://supermarket.chef.io/become-a-contributor)

**Comment by thommay** [comment]
@chuhn will you be able to sign the CLA?

**Comment by chuhn** [comment]
I guess that our legal department would have to look closely at your CLA.

As written earlier I basically cut and pasted the patch from chef/chef@5713a00 so you rather need a CLA from @btm

**Pull request closed without merging**